From 4a5ded207743421a23652604c14a5a1a1e2e6162 Mon Sep 17 00:00:00 2001
From: Jeffy Chen <jeffy.chen@rock-chips.com>
Date: Tue, 19 Nov 2019 14:46:51 +0800
Subject: [PATCH 12/81] glamor: Support using highp as default precision for
 GLES

Add new option:
--enable-gles-highp     Use highp as default precision for GLES(default:
                        no)

NOTE: Highp is optional, should only be enabled when supported by the
hardware.

Signed-off-by: Jeffy Chen <jeffy.chen@rock-chips.com>
---
 configure.ac                       | 5 +++++
 glamor/glamor_priv.h               | 5 +++++
 hw/kdrive/ephyr/ephyr_glamor_glx.c | 4 ++++
 include/dix-config.h.in            | 3 +++
 4 files changed, 17 insertions(+)

diff --git a/configure.ac b/configure.ac
index 3a32d6994..b7c9316b8 100644
--- a/configure.ac
+++ b/configure.ac
@@ -584,6 +584,7 @@ dnl kdrive options
 AC_ARG_ENABLE(libunwind,      AS_HELP_STRING([--enable-libunwind], [Use libunwind for backtracing (default: auto)]), [LIBUNWIND="$enableval"], [LIBUNWIND="auto"])
 AC_ARG_ENABLE(xshmfence,      AS_HELP_STRING([--disable-xshmfence], [Disable xshmfence (default: auto)]), [XSHMFENCE="$enableval"], [XSHMFENCE="auto"])
 
+AC_ARG_ENABLE(gles-highp,    	      AS_HELP_STRING([--enable-gles-highp], [Use highp as default precision for GLES(default: no)]), [GLES_HIGHP=$enableval], [GLES_HIGHP=no])
 
 dnl chown/chmod to be setuid root as part of build
 dnl Replaces InstallXserverSetUID in imake
@@ -2009,6 +2010,10 @@ AM_CONDITIONAL([XORG_BUS_PLATFORM], [test "x$CONFIG_UDEV_KMS" = xyes])
 AM_CONDITIONAL([XORG_DRIVER_MODESETTING], [test "x$XORG_DRIVER_MODESETTING" = xyes])
 AM_CONDITIONAL([XORG_DRIVER_INPUT_INPUTTEST], [test "x$XORG_DRIVER_INPUT_INPUTTEST" = xyes])
 
+if test "x$GLES_HIGHP" = xyes; then
+	AC_DEFINE(GLES_USE_HIGHP, 1, [Use highp as default precision for GLES])
+fi
+
 dnl glamor
 if test "x$GLAMOR" = xauto; then
 	if echo "$XORG" "$XEPHYR" | grep -q yes ; then
diff --git a/glamor/glamor_priv.h b/glamor/glamor_priv.h
index 7b144f01c..37767a5db 100644
--- a/glamor/glamor_priv.h
+++ b/glamor/glamor_priv.h
@@ -44,10 +44,15 @@
 #include <epoxy/egl.h>
 #endif
 
+#ifdef GLES_USE_HIGHP
+#define GLAMOR_DEFAULT_PRECISION \
+    "precision highp float;\n"
+#else
 #define GLAMOR_DEFAULT_PRECISION  \
     "#ifdef GL_ES\n"              \
     "precision mediump float;\n"  \
     "#endif\n"
+#endif
 
 #include "glyphstr.h"
 
diff --git a/hw/kdrive/ephyr/ephyr_glamor_glx.c b/hw/kdrive/ephyr/ephyr_glamor_glx.c
index 40b80cbe7..26feec8db 100644
--- a/hw/kdrive/ephyr/ephyr_glamor_glx.c
+++ b/hw/kdrive/ephyr/ephyr_glamor_glx.c
@@ -154,7 +154,11 @@ ephyr_glamor_setup_texturing_shader(struct ephyr_glamor *glamor)
 
     const char *fs_source =
         "#ifdef GL_ES\n"
+#ifdef GLES_USE_HIGHP
+        "precision highp float;\n"
+#else
         "precision mediump float;\n"
+#endif
         "#endif\n"
         "\n"
         "varying vec2 t;\n"
diff --git a/include/dix-config.h.in b/include/dix-config.h.in
index eed8b357c..c98923adb 100644
--- a/include/dix-config.h.in
+++ b/include/dix-config.h.in
@@ -461,6 +461,9 @@
 /* Ask fontsproto to make font path element names const */
 #define FONT_PATH_ELEMENT_NAME_CONST    1
 
+/* Use highp as default precision for GLES */
+#undef GLES_USE_HIGHP
+
 /* Build GLAMOR */
 #undef GLAMOR
 
-- 
2.38.5

